File: 2038-nearest-exit-from-entrance-in-maze/2038-nearest-exit-from-entrance-in-maze.java
class Solution {
    public int nearestExit(char[][] maze, int[] entrance) {
        int m = maze.length; 
        int n = maze[0].length;

        int[][] dis = new int[m][n];

        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                dis[i][j] = Integer.MAX_VALUE;
            }
        }

        Queue<int[]> q = new LinkedList<>();

        int[][] DIRECTIONS = new int[][]{{0,1}, {1,0}, {-1,0}, {0,-1}};
        q.offer(new int[]{entrance[0], entrance[1]});
        dis[entrance[0]][entrance[1]] = 0;
        maze[entrance[0]][entrance[1]] = '+';

        while(!q.isEmpty()){
            int[] curr = q.poll();
            int R = curr[0];
            int C = curr[1];

            for(int[] dir : DIRECTIONS){
                int nR = R + dir[0];
                int nC = C + dir[1];

                if(nR<0 || nR>=m || nC<0 || nC>=n || maze[nR][nC]=='+'){
                    continue;
                }

                if(dis[R][C] + 1 < dis[nR][nC]){
                    dis[nR][nC] = dis[R][C] + 1;
                    q.offer(new int[]{nR, nC});
                }
            }
        }

        int res = Integer.MAX_VALUE;

        
        for (int i = 0; i < n; i++) {
            if (dis[0][i] != 0) {
                res = Math.min(res, dis[0][i]);  
            }
            if (dis[m-1][i] != 0) {
                res = Math.min(res, dis[m-1][i]);  
            }
        }

        
        for (int i = 0; i < m; i++) {
            if (dis[i][0] != 0) {
                res = Math.min(res, dis[i][0]);  
            }
            if (dis[i][n-1] != 0) {
                res = Math.min(res, dis[i][n-1]);  
            }
        }

        
        return res == Integer.MAX_VALUE ? -1 : res;

    }
}

File: 2038-nearest-exit-from-entrance-in-maze/2038-nearest-exit-from-entrance-in-maze.py
class Solution(object):
    def nearestExit(self, maze, entrance):
        m = len(maze)
        n = len(maze[0])
        
        dis = [[float('inf') for _ in range(n)] for _ in range(m)]

        q = deque()
        direction = [[0,1], [0,-1], [-1,0], [1,0]]
        q.append(entrance)
        dis[entrance[0]][entrance[1]] = 0
        maze[entrance[0]][entrance[1]] = '+'

        while q:
            row, col = q.popleft()

            for dr, dc in direction:
                r = row + dr
                c = col + dc

                if 0 <= r < m and 0 <= c < n and maze[r][c] == '.':
                    dis[r][c] = dis[row][col] + 1
                    maze[r][c] = '+'  # mark as visited
                    q.append([r, c])

        res = float('inf')

        for i in range(n):
            if dis[0][i] != float('inf') and [0, i] != entrance:
                res = min(res, dis[0][i])
            if dis[m-1][i] != float('inf') and [m-1, i] != entrance:
                res = min(res, dis[m-1][i])

        for i in range(m):
            if dis[i][0] != float('inf') and [i, 0] != entrance:
                res = min(res, dis[i][0])
            if dis[i][n-1] != float('inf') and [i, n-1] != entrance:
                res = min(res, dis[i][n-1])

        return res if res != float('inf') else -1

        